Feat: Add JobName to ScrapeConfig CRD

mviswanathsai opened the pull request:
## Description

This PR will allow the user to specify the Job label they would like to appear on metrics scraped with the given ScrapeConfig CR. The PR was originally authored by @janluak. 

However, it might be of value to note that the generated scrape config does **not** populate the field `job_name`. Due to implementation details (uniqueness of job_name field), this PR achieves the desired labeling on the metrics by adopting a different approach with relabeling rules. The discussions of which can be seen in #6153.


## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [x] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Verification
<!-- How you tested it? How do you know it works? -->
Please check the [Prometheus-Operator testing guidelines](../TESTING.md) for recommendations about automated tests.

## Changelog entry

_Please put a one-line changelog entry below. This will be copied to the changelog file during the release process._

<!-- 
Your release note should be written in clear and straightforward sentences. Most often, users aren't familiar with
the technical details of your PR, so consider what they need to know when you write your release note.

Some brief examples of release notes:
- Add metadataConfig field to the Prometheus CRD for configuring how remote-write sends metadata information.
- Generate correct scraping configuration for Probes with empty or unset module parameter.
-->

```
Add JobName field to ScrapeConfig CRD

```

simonpasquier commented on pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:
```diff
@@ -118,6 +118,9 @@ func (l *ScrapeConfigList) DeepCopyObject() runtime.Object {
 // ScrapeConfigSpec is a specification of the desired configuration for a scrape configuration.
 // +k8s:openapi-gen=true
 type ScrapeConfigSpec struct {
+	// The job name assigned to scraped metrics by default.
```
```suggestion
	// The value of the `job` label assigned to the scraped metrics by default.
```

simonpasquier commented on pkg/prometheus/promcfg.go:
```diff
@@ -2577,6 +2586,10 @@ func (cg *ConfigGenerator) generateScrapeConfig(
 		cfg = cg.WithMinimumVersion("2.49.0").AppendMapItem(cfg, "scrape_protocols", sc.Spec.ScrapeProtocols)
 	}
 
+	if sc.Spec.JobName != nil || sc.Spec.RelabelConfigs != nil {
```
not sure why this new block is needed?

mviswanathsai replied:
I must've overlooked it from the original commits.

simonpasquier commented on pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:
```diff
@@ -118,6 +118,9 @@ func (l *ScrapeConfigList) DeepCopyObject() runtime.Object {
 type ScrapeConfigSpec struct {
+	// The job name assigned to scraped metrics by default.
+	// +optional
+	JobName *string `json:"jobName,omitempty"`
```
Let's add a validation that the string is at least 1 char if defined (e.g. empty isn't allowed).

mviswanathsai replied:
Added // +kubebuilder:validation:MinLength=1

ArthurSens commented on pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:
```diff
@@ -118,6 +118,10 @@ func (l *ScrapeConfigList) DeepCopyObject() runtime.Object {
 // ScrapeConfigSpec is a specification of the desired configuration for a scrape configuration.
 // +k8s:openapi-gen=true
 type ScrapeConfigSpec struct {
+	// The value of the `job` label assigned to the scraped metrics by default.
+	// +optional
+	// +kubebuilder:validation:MinLength=1
+	JobName *string `json:"jobName,omitempty"`
```
What do you think about mentioning that the job label is configured via relabelings? I think people might get different expectations if this is not documented somewhere 🤔

mviswanathsai replied:
I agree, at the first glance one might think otherwise. Would be valuable to document it as well.

simonpasquier replied:
I agree that additional details wouldn't hurt. It should also be noted that it won't change the `job_name` of the Prometheus scrape configuration which is always controlled by the operator to avoid duplicate job names which Prometheus doesn't support.

simonpasquier commented on pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:
```diff
@@ -118,6 +118,16 @@ func (l *ScrapeConfigList) DeepCopyObject() runtime.Object {
 type ScrapeConfigSpec struct {
+	// The value of the `job` label assigned to the scraped metrics by default.
+	//
+	// Note: This field does not populate the `job_name` field of the generated configuration.
+	// The `job_name` field is always controlled by the operator to prevent duplicate job names,
+	// which Prometheus does not support. Instead, all the metrics from this scrape job are
+	// labelled with the JobName by means of relabeling configs.
```
```suggestion
	// The `job_name` field in the rendered scrape configuration is always controlled by the operator to prevent duplicate job names,
	// which Prometheus does not allow. Instead the `job` label is set
	// by means of relabeling configs.
```

mviswanathsai commented:
I squashed all the commits, does the final commit authorship look good? I added myself as a co-author using commit message conventions since, after squashing down @janluak was shown as the sole author.

simonpasquier reviewed (approved):
Thanks!

mviswanathsai commented:
Could you re-run the flaky tests? @simonpasquier
